File: LeetCode/701. 二叉搜索树中的插入操作.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function insertIntoBST(root: TreeNode | null, val: number): TreeNode | null {
    const node=new TreeNode(val);
    if(!root) return node;
    let cur=root;
    while(cur){
        if(node.val<cur.val){//向左放
            if(!cur.left){
                cur.left=node;
                return root;
            }
            cur=cur.left;
            continue;
        }
        // 向右放
        if(!cur.right){
            cur.right=node;
            return root;
        }
        cur=cur.right;
    }
    return root;
};

File: LeetCode/701. 二叉搜索树中的插入操作.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn insert_into_bst(root: Option<Rc<RefCell<TreeNode>>>, val: i32) -> Option<Rc<RefCell<TreeNode>>> {
        match root {
            None => Some(Rc::new(RefCell::new(TreeNode::new(val)))),
            Some(node) => {
                let mut n = node.borrow_mut();
                if n.val > val {
                    n.left = Solution::insert_into_bst(n.left.clone(), val);
                }
                if n.val < val {
                    n.right = Solution::insert_into_bst(n.right.clone(), val);
                }
                Some(node.clone())
            }
        }
    }
}
